src/safeds_runner/server/pipeline_manager.py

```diff
@@ -0,0 +1,289 @@
+        pipeline_finder.attach()
+        main_module = f"gen_{self.pipeline.main.module}_{self.pipeline.main.pipeline}"
+        # Populate current_pipeline global, so child process can save placeholders in correct location
+        globals()["current_pipeline"] = self
+        try:
+            runpy.run_module(
+                (
+                    main_module
+                    if len(self.pipeline.main.modulepath) == 0
+                    else f"{self.pipeline.main.modulepath}.{main_module}"
+                ),
+                run_name="__main__",
+                alter_sys=True,
+            )
+            self._send_message(message_type_runtime_progress, create_runtime_progress_done())
+        except BaseException as error:  # noqa: BLE001
+            self._send_exception(error)
+        finally:
+            pipeline_finder.detach()
+
+    def execute(self) -> None:
+        """
+        Execute this pipeline in a newly created process.
+
+        Results, progress and errors are communicated back to the main process.
+        """
+        self.process.start()
+
+
+# Pipeline process object visible in child process
+current_pipeline: PipelineProcess | None = None
+
+
+def runner_save_placeholder(placeholder_name: str, value: Any) -> None:
+    """
+    Save a placeholder for the current running pipeline.
+
+    Parameters
+    ----------
+    placeholder_name : str
+        Name of the placeholder.
+    value : Any
+        Actual value of the placeholder.
+    """
+    if current_pipeline is not None:
+        current_pipeline.save_placeholder(placeholder_name, value)
+
+
+def get_backtrace_info(error: BaseException) -> list[dict[str, Any]]:
+    """
+    Create a simplified backtrace from an exception.
+
+    Parameters
+    ----------
+    error : BaseException
+        Caught exception.
+
+    Returns
+    -------
+    list[dict[str, Any]]
+        List containing file and line information for each stack frame.
+    """
+    backtrace_list = []
+    for frame in stack_data.core.FrameInfo.stack_data(error.__traceback__):
+        backtrace_list.append({"file": frame.filename, "line": int(frame.lineno)})
+    return backtrace_list
+
+
+def _get_placeholder_type(value: Any) -> str:
+    """
+    Convert a python object to a Safe-DS type.
+
+    Parameters
+    ----------
+    value : Any
+        A python object.
+
+    Returns
+    -------
+    str
+        Safe-DS name corresponding to the given python object instance.
+    """
+    if isinstance(value, bool):
+        return "Boolean"
+    if isinstance(value, float):
+        return "Float"
+    if isinstance(value, int):
+        return "Int"
+    if isinstance(value, str):
+        return "String"
+    if isinstance(value, object):
+        object_name = type(value).__name__
+        if object_name == "function":
+            return "Callable"
+        if object_name == "NoneType":
+            return "Null"
+        return object_name
```

Comment:
Can also be expressed using [pattern matching](https://stackoverflow.com/a/67524642). It's up to you whether you want to change it or leave it.